File: cloud-collect-core/src/main/java/com/windea/demo/cloudcollect/core/repository/CollectRepository.java
package com.windea.demo.cloudcollect.core.repository;

import com.windea.demo.cloudcollect.core.domain.entity.Collect;
import com.windea.demo.cloudcollect.core.domain.enums.CollectType;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;

import java.util.Optional;

public interface CollectRepository extends JpaRepository<Collect, Long> {
	Optional<Collect> findByName(String name);

	Page<Collect> findByUser_IdAndDeleteStatus(Long userId, Boolean deleteStatus, Pageable pageable);

	long countByUser_IdAndDeleteStatus(Long userId, Boolean deleteStatus);

	Page<Collect> findByUser_IdAndNameContainsAndDeleteStatusFalse(Long userId, String name, Pageable pageable);

	Page<Collect> findByCategory_IdAndDeleteStatusFalse(Long categoryId, Pageable pageable);

	long countByCategory_IdAndDeleteStatusFalse(Long categoryId);

	@Query("from Collect c, in(c.tags) t where t.id=:tagId and c.deleteStatus=false")
	Page<Collect> findByTag_IdAndDeleteStatusFalse(Long tagId, Pageable pageable);

	@Query("select count(c) from Collect c, in(c.tags) t where t.id=:tagId and c.deleteStatus=false")
	long countByTag_IdAndDeleteStatusFalse(Long tagId);

	Page<Collect> findByUser_IdAndTypeAndDeleteStatusFalse(Long userId, CollectType type, Pageable pageable);

	long countByUser_IdAndTypeAndDeleteStatusFalse(Long userId, CollectType type);

	@Query("from Collect c, in(c.praiseByUserList) u where u.id=:praiseByUserId and c.deleteStatus=false")
	Page<Collect> findByPraiseByUser_IdAndDeleteStatusFalse(Long praiseByUserId, Pageable pageable);

	@Query("select count(c) from Collect c, in(c.praiseByUserList) u where u.id=:praiseByUserId and c" +
		".deleteStatus=false")
	long countByPraiseByUser_IdAndDeleteStatusFalse(Long praiseByUserId);

	Page<Collect> findByNameContainsAndDeleteStatusFalse(String name, Pageable pageable);

	boolean existsByUser_IdAndName(Long userId, String name);
}

File: cloud-collect-core/src/main/kotlin/com/windea/demo/cloudcollect/core/repository/CollectRepository.kt
package com.windea.demo.cloudcollect.core.repository

import com.windea.demo.cloudcollect.core.domain.entity.*
import com.windea.demo.cloudcollect.core.enums.*
import org.springframework.data.domain.*
import org.springframework.data.jpa.repository.*

interface CollectRepository : JpaRepository<Collect, Long> {
	fun findByNameAndUserId(name: String, userId: Long): Collect?
	
	fun findAllByNameContains(name: String, pageable: Pageable): Page<Collect>
	
	fun findAllByCategoryId(categoryId: Long, pageable: Pageable): Page<Collect>
	
	fun findAllByCategoryName(categoryName: String, pageable: Pageable): Page<Collect>
	
	fun findAllByCategoryNameContains(categoryName: String, pageable: Pageable): Page<Collect>
	
	fun findAllByTagsId(tagId: Long, pageable: Pageable): Page<Collect>
	
	fun findAllByTagsName(tagName: String, pageable: Pageable): Page<Collect>
	
	fun findAllByTagsNameContains(tagName: String, pageable: Pageable): Page<Collect>
	
	fun findAllByUserId(userId: Long, pageable: Pageable): Page<Collect>
	
	fun findAllByNameContainsAndUserId(name: String, userId: Long, pageable: Pageable): Page<Collect>
	
	fun findAllByCategoryNameContainsAndUserId(categoryName: String, userId: Long, pageable: Pageable): Page<Collect>
	
	fun findAllByTagsNameContainsAndUserId(tagName: String, userId: Long, pageable: Pageable): Page<Collect>
	
	fun findAllByTypeAndUserId(type: CollectType, userId: Long, pageable: Pageable): Page<Collect>
	
	fun findAllByPraiseByUsersId(praiseByUserId: Long, pageable: Pageable): Page<Collect>
	
	fun findAllByNameContainsAndPraiseByUsersId(name: String, praiseByUserId: Long, pageable: Pageable): Page<Collect>
	
	fun findAllByCategoryNameContainsAndPraiseByUsersId(categoryName: String, praiseByUserId: Long, pageable: Pageable): Page<Collect>
	
	fun findAllByTagsNameContainsAndPraiseByUsersId(tagName: String, praiseByUserId: Long, pageable: Pageable): Page<Collect>
	
	fun findAllByTypeContainsAndPraiseByUsersId(type: CollectType, praiseByUserId: Long, pageable: Pageable): Page<Collect>
	
	fun countByCategoryId(categoryId: Long): Long
	
	fun countByTagsId(tagId: Long): Long
	
	fun countByUserId(userId: Long): Long
	
	fun countByPraiseByUsersId(praiseByUserId: Long): Long
	
	fun existsByNameAndUser(name: String, user: User): Boolean
	
	fun existsByIdAndPraiseByUsers(id: Long, praiseByUser: User): Boolean
}
